### DISCUSSION TITLE
Relay-Pagination cancels second query and sends "teardown" despite partial is true

### DISCUSSION THREAD

**Discussion by psteinroe** [urql-graphql/urql#1965] [Q&A] [answered]
Hi,

I am implementing a custom relay-pagination that works with nested pagination args. You can find the code here:
https://gist.github.com/steinroe/62f9766c1e03e8bca78cd5ce2314f7e5

All tests are passing, but it does not properly paginate. Here is the setup:

```typescript
export const createUrqlClient = (
  endpointURL: string,
  subscriptionURL: string,
  fetchAuthState: () => Promise<AuthState | null>,
  onGqlError: (error: GraphQLError) => void,
  onNetworkError: (error: Error) => void
) => {
  const wsClient = createWSClient({
    url: subscriptionURL,
    connectionParams: async () => {
      const idToken = (await fetchAuthState())?.idToken;
      if (!idToken) return {};
      return {
        authorization: `Bearer ${idToken}`,
      };
    },
  });

  return createClient({
    url: endpointURL,
    exchanges: [
      devtoolsExchange,
      dedupExchange,
      cacheExchange({
        schema: introspectionResult as IntrospectionData,
        resolvers: {
          Query: {
            invitees: relayPagination(),
            businessPartners: relayPagination(),
            realEstates: relayPagination(),
          },
        },
      }),
      errorExchange({
        onError: ({ networkError, graphQLErrors }) => {
          if (networkError) onNetworkError(networkError);
          if (graphQLErrors) {
            graphQLErrors.map((e) => onGqlError(e));
          }
        },
      }),
      fetchOptionsExchange(async (fetchOptions) => {
        const auth = await fetchAuthState();
        if (!auth) return fetchOptions;
        return Promise.resolve({
          ...fetchOptions,
          headers: {
            Authorization: `Bearer ${auth.idToken}`,
          },
        });
      }),
      fetchExchange,
      subscriptionExchange({
        forwardSubscription: (operation) => ({
          subscribe: (sink) => ({
            unsubscribe: wsClient.subscribe(operation, sink),
          }),
        }),
      }),
    ],
  });
};
```

When I now use a query, e.g. 
```
query realEstates($paging: CursorPaging){
  realEstates(paging: $paging) {
    edges{
      node{
        id
      }
    }
    pageInfo{
      hasNextPage
      startCursor
      endCursor
      hasPreviousPage
    }
  }
}
```
with
```typescript
  const [cursor, setCursor] = useState();

  const [{ data, fetching, stale }] = useRealEstatesQuery({
    variables: { paging: { first: PAGE_SIZE, after: cursor } },
    requestPolicy: 'cache-and-network',
  });
```
it seems to resolve correctly by merging the first page and setting the partial flag to true, since the second page cannot be found. However, the operation is not forwarded. Instead, it returns the first page, writes that one to the cache but with the args for the second page and emits a teardown operation to the following exchanges:
```
context: {url: 'xxx', fetchOptions: undefined, fetch: undefined, preferGetMethod: false, requestPolicy: 'cache-and-network', …}
key: 1245865492
kind: "teardown"
query: {kind: 'Document', definitions: Array(1), loc: Location, __key: 433463296}
variables:
paging:
after: "eyJ0eXBlIjoia2V5c2V0IiwiZmllbGRzIjpbeyJmaWVsZCI6ImlkIiwidmFsdWUiOiJhYTU5MGYxNy05N2U2LTQ0YmEtODNjOC03YzNhMzg2Zjg3ODMifV19"
first: 5
[[Prototype]]: Object
```
The devtools show the second-page query with the data of the first page 
![Screenshot 2021-09-15 at 17 00 05](https://user-images.githubusercontent.com/19429600/133458338-dd2924fc-7d1a-49b3-a367-51151147eea2.png)
and the operation for the second page is continuing indefinitely.
![Screenshot 2021-09-15 at 17 01 04](https://user-images.githubusercontent.com/19429600/133458481-a80e0570-f38e-4458-8692-070e77d2146f.png)

My expectation would be that since partial is true, it executes the query for the second page.

**Comment by psteinroe** [comment, marked as answer]
Just noticed that this was a bug:
https://github.com/FormidableLabs/urql/discussions/1965

and it is fixed with the current version.

**Reply by kitten** [reply]
So sorry! I suspected this when I had a one minute glance at your issue here, but didn't reply then since I'm off work this week ❤️ I believe this indeed relates to an issue we fixed recently. We introduced a small regression in our feature that aims to increase referential equality (and hence in frameworks that rely heavily on it for performance optimisations) so we hope that it is still worth it even given the momentary regression we had to fix 😅 https://github.com/FormidableLabs/urql/blob/main/exchanges/graphcache/CHANGELOG.md#431
